activejob/lib/active_job/arguments.rb

```diff
@@ -93,6 +93,14 @@ def serialize_argument(argument)
         when -> (arg) { arg.respond_to?(:permitted?) }
           serialize_indifferent_hash(argument.to_h)
         else
+          if BigDecimal === argument && !ActiveJob.use_big_decimal_serializer
+            ActiveSupport::Deprecation.warn(<<~MSG)
```

Comment:
I am unsure if this will show the correct "caller location" for users. In the tests, it always showed as coming from Minitest internals, regardless of what level of `caller_locations(n)` I passed...

Reply:
Yeah, you are quite deep here so it's hard to have the proper location, but I think that is fine. 

My concern would be the error message:

  - I think it need to say when it will be removed (Rails 7.2)
  - I think it should say not to enable it on first deploy.

Reply:
I've updated it to

```
Primitive serialization of BigDecimal job arguments is deprecated as it may serialize via .to_s using certain queue adapters.
Enable config.active_job.use_big_decimal_serializer to use BigDecimalSerializer instead, which will be mandatory in Rails 7.2.

Note that if you application has multiple replicas, you should only enable this setting after successfully deploying your app to Rails 7.1 first.
This will ensure that during your deployment all replicas are capable of deserializing arguments serialized with BigDecimalSerializer.
```

Reply:
As far as `caller_locations` goes, the thing that doesn't make sense to me is that if I do

```ruby
if BigDecimal === argument && !ActiveJob.use_big_decimal_serializer
  puts
  puts caller_locations.take(10)
  puts '----'
  puts caller_locations(4).first
  puts '----'
  ActiveSupport::Deprecation.warn(<<~MSG, caller_locations(4)) # ...
```

I get

```
/Users/sambostock/src/github.com/sambostock/rails/activejob/lib/active_job/arguments.rb:35:in `block in serialize'
/Users/sambostock/src/github.com/sambostock/rails/activejob/lib/active_job/arguments.rb:35:in `map'
/Users/sambostock/src/github.com/sambostock/rails/activejob/lib/active_job/arguments.rb:35:in `serialize'
/Users/sambostock/src/github.com/sambostock/rails/activejob/test/cases/argument_serialization_test.rb:62:in `block in <class:ArgumentSerializationTest>'
/Users/sambostock/.gem/ruby/3.1.0/gems/minitest-5.16.1/lib/minitest/test.rb:98:in `block (3 levels) in run'
/Users/sambostock/.gem/ruby/3.1.0/gems/minitest-5.16.1/lib/minitest/test.rb:195:in `capture_exceptions'
/Users/sambostock/.gem/ruby/3.1.0/gems/minitest-5.16.1/lib/minitest/test.rb:95:in `block (2 levels) in run'
/Users/sambostock/.gem/ruby/3.1.0/gems/minitest-5.16.1/lib/minitest.rb:296:in `time_it'
/Users/sambostock/.gem/ruby/3.1.0/gems/minitest-5.16.1/lib/minitest/test.rb:94:in `block in run'
/Users/sambostock/.gem/ruby/3.1.0/gems/minitest-5.16.1/lib/minitest.rb:391:in `on_signal'
----
/Users/sambostock/src/github.com/sambostock/rails/activejob/test/cases/argument_serialization_test.rb:62:in `block in <class:ArgumentSerializationTest>'
----
DEPRECATION WARNING: Primitive serialization of BigDecimal job arguments is deprecated as it may serialize via .to_s using certain queue adapters.
Enable config.active_job.use_big_decimal_serializer to use BigDecimalSerializer instead, which will be mandatory in Rails 7.2.

Note that if you application has multiple replicas, you should only enable this setting after successfully deploying your app to Rails 7.1 first.
This will ensure that during your deployment all replicas are capable of deserializing arguments serialized with BigDecimalSerializer.
 (called from block (3 levels) in run at /Users/sambostock/.gem/ruby/3.1.0/gems/minitest-5.16.1/lib/minitest/test.rb:98)
```

It appears `caller_locations(4)` is the correct place to "originate" the deprecation from (at least for this test), but for some reason, no matter what I pass to `warn`, it always shows
```
  (called from block (3 levels) in run at /Users/sambostock/.gem/ruby/3.1.0/gems/minitest-5.16.1/lib/minitest/test.rb:98)
```

Reply:
Oh, I think I just answered my own question: It looks like `ActiveSupport::Deprecation` [filters the callstack to exclude Rails itself](https://github.com/rails/rails/blob/fa7f977e2d68900b6c9fdf12331c507759e443f7/activesupport/lib/active_support/deprecation/reporting.rb#L153), and since the `minitest` is the first path not within the Rails gem root, that's what we get.

I'll use one of the bug report templates to run into the deprecation from outside Rails and figure out the correct `caller_location`.